Adds KaTeX math engine via the 'katex' ruby gem

glebm opened the pull request:
This is a simpler all batteries included alternative to the `sskatex` engine.

glebm commented:
There will be test errors that I think are due to an issue in sskatex (https://github.com/ccorn/sskatex/issues/2)

ccorn commented:
I have merged the proposed fix by @glebm to SsKaTeX and made a new release (no longer needs to overwrite `EXECJS_RUNTIME` to defer autodetection). Re-triggering the CI tests should now load the `katex` gem normally.

glebm commented:
I've updated the PR to trigger a test re-run

glebm commented:
The pre-2.3 targets are failing because the `katex` gem require ruby `~> 2.3`.

@gettalong Perhaps kramdown could drop official support for older rubies?
Ruby 2.2 EOL is scheduled for 2018-03-31.

ccorn commented:
You may have to constrain the `katex` gem version in `.travis.yml` so that updated versions with a newer `katex.js` (and different HTML output) do not automatically trigger CI failures.

ccorn commented:
For earlier rubies, the load failure would be reflected in the `AVAILABLE` member. You just need to exlude the installation attempts and the new test cases when the engine is not available.

* You could replace the katex gem installation in `.travis.yml` with:

      - ruby -e 'exit(RUBY_VERSION < "2.3" ? 0 : 1)' || gem install katex

* and augment `test/test_files.rb` with `KATEX_AVAILABLE` guards like `SsKaTeX does.

glebm commented:
Thanks @ccorn, I've done as you suggested and Travis is now green

gettalong commented:
So @glebm, could you explain to me what this implementation offers over sskatex?

glebm commented:
@gettalong The `katex` gem is much easier to set up, because it bundles all the KaTeX assets and register them automatically with all the popular Ruby web frameworks (Rails, Sprockets without Rails, and Hanami).

ccorn commented:
From my point of view:

* The `katex` engine is a locked-down fixed-configuration engine for untrusted users:
  - No Javascript fragments configurable, thus suitable for applications which need to disable all means of code injection
  - JS engine determined by ExecJS autodetection, not overridable
  - Needs to cache only one execution context (within the gem) because there is only one configuration available
  - KaTeX version determined by gem version; `throwOnError` turned off; no `macros` nor other options that SsKaTeX's `katex_opts` would allow to be passed to KaTeX
  - No logging
* SsKaTeX is for trusted users who need more control or stability:
  - KaTeX version is controlled by the user directly
  - Can pass KaTeX options, throws on errors by default
  - Can change all JS engine and code aspects, supports logging/debug messages
  - May be able to interface with MathJax by using different `js_libs`
  - Makes sure to pass ASCII strings only
    * Useful for ancient engines like SpiderMonkey 1.8.5 which can run KaTeX just fine but may have been compiled without support for UTF-8 I/O
    * Note that such engines are deprecated in ExecJS and not autoselected.

Let me add that KaTeX version control is important; e.g. the current 0.9.0-beta release makes fraction lines disappear in some browsers.

ccorn commented:
In `Rakefile`, the added line `s.add_development_dependency 'katex'` may need to be conditioned on `RUBY_VERSION` as well.

To avoid repeating that criterion too often, the `KATEX_AVAILABLE` test in `test/test_files.rb` could be made similar to the one for `MATHJAX_NODE_AVAILABLE` (that is, just test the `AVAILABLE` member). This would then work regardless of whether `katex` is a development dependency or not.

gettalong commented:
@glebm If it is only that, then there is no real reason to add this to kramdown. If there were any big features that katex has but sskatex hasn't, than it would make sense. But from kramdown's point of view it doesn't make sense to provide two ways to do effectively the same.

ccorn commented:
The target user sets are disjoint: untrusted users on a github-pages-like service vs trusted individuals with their own specific setups. It solves the security issue discussed in #471 by bundling all the ingredients and shielding kramdown from the associated maintenance costs. I imagine that the `katex` version could be advertised for general use, and if individuals run into problems with their personal setups, they can be asked to use SsKaTeX for better control.

gettalong commented on test/test_files.rb:
```diff
@@ -42,14 +42,16 @@ class TestFiles < Minitest::Test
                            end
 
-  KATEX_AVAILABLE = begin
+  SSKATEX_AVAILABLE = begin
```
The following lines should be indented two additional spaces since `SSKATEX_AVAILABLE` is two characters longer than `KATEX_AVAILABLE`.

glebm replied:
Done

glebm commented on lib/kramdown/converter/math_engine/katex.rb:
```diff
@@ -0,0 +1,33 @@
+
+    def self.call(converter, el, opts)
+      display_mode = el.options[:category] == :block
+      result = ::Katex.render el.value,
```
Parenthesis are optional here. Do you have a style guide / a lint command?

gettalong replied:
No, not for kramdown. I would just use them here.

glebm replied:
Done.

glebm commented on lib/kramdown/converter.rb:
```diff
@@ -50,7 +50,7 @@ module Converter
     configurable(:math_engine)
 
-    ['Mathjax', "MathjaxNode", "SsKaTeX", "Ritex", "Itex2MML"].each do |klass_name|
+    %w(Mathjax MathjaxNode Katex SsKaTeX Ritex Itex2MML).each do |klass_name|
```
It then does not fit on one line. How do you want it indented if it's split across multiple lines?

glebm replied:
Nevermind, from looking at other code I see that the undocumented column limit is not 80.

glebm replied:
Done

gettalong replied:
Column limit is 100. When I have some more time I will add support for a linting tool...

glebm commented on .travis.yml:
```diff
@@ -22,6 +22,8 @@ before_install:
   - gem install rake
   - gem install minitest coderay stringex ritex execjs sskatex
+  - >
+    ruby -e 'exit RUBY_VERSION < "2.3" ? 0 : 1' || gem install katex -v 0.3.0
```
It contains a `:`. Do you know how it should be escaped?

glebm replied:
Figured it out, done.

gettalong replied:
Hm... always something new to learn 👍 You are correct, the colon poses a problem, sorry. So leave that as it is.

glebm commented on doc/math_engine/katex.page:
```diff
@@ -0,0 +1,29 @@
+---
```
Done

ccorn commented:
* Adding KaTeX to the Output list in `doc/documentation.page` makes no sense.
* What requires updates are `doc/converter/html.page` (math section) and `doc/documentation.template` (math engines).
* In `test/test_files.rb`, line 46 (below SSKATEX_AVAILABLE) still requires +2 indent.

ccorn commented:
The sources `doc/converter/html.page` and/or the individual math engine's pages should probably include a discussion `katex` vs `sskatex`:

| katex | sskatex |
|:-|:-|
| easy to use | more control |
| for untrusted users | for trusted users |

 I'd provide one in a follow-up PR if you do not want to touch SsKaTeX parts. In two rough sentences: Use `katex` unless you need more control and can be trusted to configure `sskatex`. Do not make the `sskatex` gem available to a service with untrusted users.

glebm commented:
@ccorn I've addressed your comments.

> I'd provide one in a follow-up PR if you do not want to touch SsKaTeX parts.

I'm going to take you up on this, thank you.

ccorn commented:
In `doc/math_engine/katex.page`, it should IMHO be mentioned where the fonts and CSS are stored, at least with a link to the gem's [manual registration](https://github.com/glebm/katex-ruby#manual-registration) section so that e.g. Jekyll users can configure their system appropriately.

glebm commented:
@ccorn I've added a link to the Assets section in the gem's docs to avoid duplicating any specifics here, because they may get out of date. For example, as Jekyll is a common use case, I'd like to support Automatic Registration for Jekyll in the future as well.

ccorn commented:
Just mentioning that there are no Rdoc comments in `lib/kramdwon/converter/math_engine/katex.rb`. But there is really not much to say anyway, and the adjacent `itex2mml.rb` or `ritex.rb` (which have very similar structure) are about as sparse.

ccorn commented:
IMHO, this PR (checked up to 4216c670e15d515c4f6f26a8c8713344eae801c0) now has all required ingredients.
(I'd move the `KaTeX` in the `documentation.template` closer to `SsKaTeX`, leaving the default `MathJax` at the top of the list, but that's just my taste.)
I have also tried it on a small Jekyll tree with a couple hundred formulas, and it works (after setting the KaTeX assets to those of v0.8.3; I sometimes try other snapshots).

gettalong commented:
@glebm Thanks for all the work - please squash the commits into one and rebase on current master, then I will merge.

Thanks for your contribution!

glebm commented:
@gettalong Done!

gettalong commented:
@glebm Thanks - I have merged the commit.
